zeus/vcs/api.py

```diff
@@ -0,0 +1,197 @@
+                )
+                return json_response({"error": "invalid_auth"}, status=401)
+
+            if "uid" in token:
+                with sentry_sdk.configure_scope() as scope:
+                    scope.user = {"id": token["uid"]}
+
+            if "repo_ids" not in token or repo_id not in token["repo_ids"]:
+                current_app.logger.debug(
+                    "vcs-server.invalid-request command=%s tenant=%s reason=invalid-repo",
+                    func.__name__,
+                    token,
+                )
+                return json_response({"error": "access_denied"}, status=400)
+
+            current_app.logger.debug(
+                "vcs-server.request repo_id=%s command=%s tenant=%s",
+                repo_id,
+                func.__name__,
+                token,
+            )
+
+            try:
+                return await func(request, repo_id, *args, **kwargs)
+            except InvalidPublicKey:
+                current_app.logger.exception(
+                    "vcs-server.invalid-pubkey repo_id=%s", repo_id
+                )
+                return json_response({"error": "invalid_pubkey"}, status=400)
+            except Exception:
+                current_app.logger.exception(
+                    "vcs-server.unhandled-error repo_id=%s", repo_id
+                )
+                return json_response({"error": "unknown_error"}, status=500)
+
+        return tmp()
+
+    return wrapper
+
+
+@span("health_check")
+@log_errors
+async def health_check(request):
+    return Response(text='{"ok": true}')
+
+
+@span("stmt.log")
+@api_request
+@log_errors
+async def stmt_log(request, repo_id):
+    vcs = await get_vcs(request.app, repo_id)
+
+    queue = request.app["queue"]
+
+    parent = request.query.get("parent")
+    branch = request.query.get("branch")
+    offset = int(request.query.get("offset") or 0)
+    limit = int(request.query.get("limit") or 100)
+
+    try:
+        log_results = vcs.log(parent=parent, branch=branch, offset=offset, limit=limit)
+    except UnknownRevision:
+        # we're running a lazy update here if it didnt already exist
+        log_results = vcs.log(
+            parent=parent,
+            branch=branch,
+            offset=offset,
+            limit=limit,
+            update_if_exists=True,
+        )
+
+    results = []
+    for revision in log_results:
+        results.append(
+            {
+                "sha": revision.sha,
+                "message": revision.message,
+                "authors": revision.get_authors(),
+                "author_date": revision.author_date.isoformat(),
+                "committer": revision.get_committer(),
+                "committer_date": (
+                    revision.committer_date or revision.author_date
+                ).isoformat(),
+                "parents": revision.parents,
+            }
+        )
+        await queue.put(["revision", {"repo_id": repo_id, "revision": revision}])
```

Comment:
this will overload a local worker quite quickly i expect, and do useless db calls

id like to put an lru cache in for "does this revision need saved", but theres nothing useful in stdlib for it